plutus-core/src/Language/PlutusCore/CBOR.hs

```diff
@@ -94,9 +94,9 @@ decodeConstructorTag = decodeWord
 
 -- See Note [The G, the Tag and the Auto].
 instance Closed uni => Serialise (Some (TypeIn uni)) where
-    encode (Some (TypeIn uni)) = encodeConstructorTag . fromIntegral $ tagOf uni
+    encode (Some (TypeIn uni)) = encode . map (fromIntegral :: Int -> Word) $ encodeUni uni
```

Comment:
This will make encoding constants a bit more expensive, but I think that's fine.

Reply:
If we really care about  saving the overhead incurred by encoding a list, can't we encode a single `Int` if the list has length 1 (which will always be the case at present) and encode the whole list otherwise, and then use `deserialiseOrFail` to deal with the two possibilities during decoding? Of course that'll make decoding a bit more expensive instead.  (Also, CBOR's probably going away.)

Reply:
(I'm not sure why that was `encodeConstructorTag` there: that's suppose to be used to encode Haskell constructors.)

Reply:
> If we really care about saving the overhead incurred by encoding a list, can't we encode a single Int if the list has length 1

Would that be more optimal? It seems like the overhead of distinguish between this two cases is not very much different from just serializing the length of the list (which is what CBOR does, I believe).

Reply:
Yeah, I think this is probably not too bad. And there are comparatively few constants in a program compared to other AST nodes.

Reply:
> It seems like the overhead of distinguish between this two cases is not very much different from just serializing the length of the list (which is what CBOR does, I believe).

Just for the sake of nitpicking, that's not actually what CBOR does (or more precisely, what CBORG does).  For a stream of items like a list or array, CBOR allows you to do two different things. You can emit a tag saying "here's a list of n items" and follow it with the encodings of the individual items; if the length is small you can fit it and the beginning-of-list marker into a single byte.  The encodings for small integers all fit into one byte as well, so you could encode `[2]` using two bytes.  However, you can also emit a tag saying "here's the start of a list of indefinite length", emit the encoding of the items, and then emit a tag saying "end of list", so `[2]` would require 3 bytes.  For nonempty lists, CBORG uses the latter strategy (or at least I'm pretty sure it does).  

So my suggestion would require 1 byte in to encode the common case of a one-element list with a single small entry, whereas encoding the whole list would require 3 bytes.  So if you had ten constants in a program you could save up to a massive 20 bytes!

As Michael says, we're probably not going to have many constants, so it's probably not worth the trouble.  If we actually cared (I don't) we could always do a before and after comparison to see if it really makes any difference, since I think my suggestion with the scheme in this PR will produce exactly the same encoding as the previous scheme for all of our existing examples.

[Back to my so-called holiday now...]

Reply:
@kwxm thank you, I just quickly looked on wiki what CBOR does, but did not look into the exact details of `cborg`.

Yeah, 3 bytes is definitely worse than 1 byte. Maybe we should do some actual measurements and adopt your strategy.

Reply:
Note that we're almost certainly going to change our binary serialization format soon, so don't waste too much thought on CBOR.